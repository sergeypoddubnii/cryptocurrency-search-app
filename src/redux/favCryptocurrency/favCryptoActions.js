import favCyptoTypes from './favCyptoTypes';

const addToFavorite = crypto => {
  return {
    type: favCyptoTypes.ADD_TO_FAVORITE,
    payload: { crypto },
  };
};

const removeFromFavorite = id => {
  return {
    type: favCyptoTypes.REMOVE_FROM_FAVORITE,
    payload: { id },
  };
};

export default {
  addToFavorite,
  removeFromFavorite,
};
