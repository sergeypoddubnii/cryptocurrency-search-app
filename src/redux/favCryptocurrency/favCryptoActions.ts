import {
  ADD_TO_FAVORITE,
  REMOVE_FROM_FAVORITE,
  CryptoCurrency,
  AddToFavoriteAction,
  RemoveFromFavoriteAction,
} from './favCyptoTypes';

const addToFavorite = (crypto: CryptoCurrency): AddToFavoriteAction => {
  return {
    type: ADD_TO_FAVORITE,
    payload: { crypto },
  };
};

const removeFromFavorite = (id: string): RemoveFromFavoriteAction => {
  return {
    type: REMOVE_FROM_FAVORITE,
    payload: { id },
  };
};

export default {
  addToFavorite,
  removeFromFavorite,
};
