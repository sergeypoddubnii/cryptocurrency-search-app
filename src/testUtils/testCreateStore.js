import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const testCreateStore = (store = {}) => configureStore([thunk])({ ...store });

export default testCreateStore;
