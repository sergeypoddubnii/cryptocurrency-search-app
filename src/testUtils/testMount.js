import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

const testMount = (Component, store = {}) => mount(<Provider store={store}>{Component}</Provider>);

export default testMount;
