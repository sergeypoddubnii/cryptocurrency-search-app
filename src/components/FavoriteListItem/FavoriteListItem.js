import React from 'react';
import { useDispatch } from 'react-redux';
import favCryptoActions from '../../redux/favCryptocurrency/favCryptoActions';
import styled from 'styled-components';

const Li = styled.li`
  list-style-type: none;
  margin-bottom: 5px;
`;

const Button = styled.button`
  background: #f00;
  border-radius: 50%;
  font-size: 18px;
  line-height: 18px;
  width: 18px;
  padding: 0;
  border: none;
  outline: none;
`;

const Title = styled.h2`
  font-size: 18px;
  padding: 0;
  margin: 0;
`;

const FavoriteListItem = ({ name, symbol, price, id }) => {
  const dispatch = useDispatch();

  const removeFromFavoriteHandler = () =>
    dispatch(favCryptoActions.removeFromFavorite(id));

  return (
    <Li>
      <Title>
        {name}
        &nbsp;
        {symbol}
      </Title>
      <p>
        Price:{price.toFixed(2)}$ &nbsp;
        <Button
          onClick={removeFromFavoriteHandler}
          type="button"
          title="remove from favorite list"
        >
          -
        </Button>
      </p>
    </Li>
  );
};

export default FavoriteListItem;
