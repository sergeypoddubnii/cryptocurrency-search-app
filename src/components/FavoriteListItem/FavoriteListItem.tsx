import React, { FC } from 'react';
import { useDispatch } from 'react-redux';
import favCryptoActions from '../../redux/favCryptocurrency/favCryptoActions';
import styled from 'styled-components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTrashAlt } from '@fortawesome/free-solid-svg-icons';
import withWrapperListItem from '../../hoc/withWrapperListItem';

const Button = styled.button`
  background: transparent;
  color: red;
  font-size: 14px;
  line-height: 18px;
  width: 18px;
  padding: 0;
  border: none;
  outline: none;
`;

interface FavoriteListItemProps {
  id: string;
}

const FavoriteListItem: FC<FavoriteListItemProps> = ({ id }) => {
  const dispatch = useDispatch();

  const removeFromFavoriteHandler = () =>
    dispatch(favCryptoActions.removeFromFavorite(id));

  return (
    <Button
      onClick={removeFromFavoriteHandler}
      type="button"
      title="remove from favorite list"
    >
      <FontAwesomeIcon icon={faTrashAlt} />
    </Button>
  );
};

export default withWrapperListItem(FavoriteListItem);
