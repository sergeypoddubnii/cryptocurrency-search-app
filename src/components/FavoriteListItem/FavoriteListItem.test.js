import React from 'react';
import { shallow } from 'enzyme';
import FavoriteListItem from './FavoriteListItem.tsx';
import testCreateStore from '../../testUtils/testCreateStore';
import testMount from '../../testUtils/testMount';
import favCryptoActions from '../../redux/favCryptocurrency/favCryptoActions.ts';

describe('FavoriteListItem', () => {
	const favoriteListItem = shallow(<FavoriteListItem />);
	it('should render properly', () => {
		expect(favoriteListItem).toMatchSnapshot();
	});
	it('should dispatch the correct action on removeFromFavoriteHandler click', () => {
		const store = testCreateStore();
		store.dispatch = jest.fn();
		const favoriteListItemMount = testMount(<FavoriteListItem />, store);

		favoriteListItemMount.find('button').simulate('click');
		expect(store.dispatch).toHaveBeenCalledWith(
			favCryptoActions.removeFromFavorite(),
		);
	});
});
