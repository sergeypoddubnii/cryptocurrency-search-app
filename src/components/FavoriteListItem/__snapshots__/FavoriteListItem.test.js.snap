// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteListItem should render properly 1`] = `
<styled.li>
  <styled.h2>
    Â 
  </styled.h2>
  <div>
    <span>
      $
    </span>
    <FavoriteListItem />
  </div>
</styled.li>
`;
